desc/grid.py

```diff
@@ -33,36 +33,68 @@ class Grid(IOAble):
         "_weights",
         "_axis",
         "_node_pattern",
+        "_unique_rho_idx",
+        "_unique_theta_idx",
+        "_unique_zeta_idx",
+        "_inverse_rho_idx",
+        "_inverse_theta_idx",
+        "_inverse_zeta_idx",
         "_num_rho",
         "_num_theta",
         "_num_zeta",
     ]
 
     def __init__(self, nodes, sort=True):
-
-        nodes = np.atleast_2d(nodes)
-        self._L = np.unique(nodes[:, 0]).size
-        self._M = np.unique(nodes[:, 1]).size
-        self._N = np.unique(nodes[:, 2]).size
         self._NFP = 1
         self._sym = False
         self._node_pattern = "custom"
 
         self._nodes, self._spacing = self._create_nodes(nodes)
 
-        self._enforce_symmetry()
+        dtheta_scale = self._enforce_symmetry()
         if sort:
             self._sort_nodes()
         self._find_axis()
         self._count_nodes()
-        self._scale_weights()
+        self._scale_weights(dtheta_scale)
 
     def _enforce_symmetry(self):
-        """Enforce stellarator symmetry."""
-        if self.sym:  # remove nodes with theta > pi
+        """Enforce stellarator symmetry.
+
+        1. Remove nodes with theta > pi.
+        2. Rescale theta spacing to preserve dtheta weight.
+            Need to rescale on each theta coordinate curve by a different factor.
+            dtheta should = 2pi / number of nodes remaining on that theta curve
+
+        Returns
+        -------
+        dtheta_scale : ndarray
+            The multiplicative factor to scale the theta spacing for each theta curve.
+                number of nodes / (number of nodes - number of nodes to delete)
+
+        """
+        if self.sym:
             non_sym_idx = np.where(self.nodes[:, 1] > np.pi)
+            __, inverse, nodes_per_rho_surf = np.unique(
+                self.nodes[:, 0], return_inverse=True, return_counts=True
+            )
+            __, non_sym_per_rho_surf = np.unique(
+                self.nodes[non_sym_idx, 0], return_counts=True
+            )
+            if len(nodes_per_rho_surf) > len(non_sym_per_rho_surf):
+                # edge case where surface closest to axis lacks theta > pi nodes
+                pad_count = len(nodes_per_rho_surf) - len(non_sym_per_rho_surf)
+                non_sym_per_rho_surf = np.pad(non_sym_per_rho_surf, (pad_count, 0))
```

Comment:
`nodes_per_rho_surf` is the total number of nodes on each surface, and has a length equal to `grid.num_rho`. 
`non_sym_per_rho_surf` is the number of "non-symmetric" nodes on each surface, or the number of nodes with theta > pi on each surface. It's length is _normally_ also equal to `grid.num_rho`. 
The edge case is with a `ConcentricGrid` where the innermost surface only has a single node at theta=0. In this case, `non_sym_per_rho_surf` has a length of `grid.num_rho - 1` because it doesn't have a value for that innermost surface. And in those cases, we pad the array with a 0 since there are "0 non-symmetric nodes" on that surface.

Reply:
That’s right but also on high resolution there are multiple rho surfaces without theta > pi nodes.

Reply:
Feel free to add Daniel's longer explanation to the comment in the code if it is clarifying.